source/agora/common/EnrollmentManager.d

```diff
@@ -0,0 +1,407 @@
+
+mixin AddLogger!();
+
+/*******************************************************************************
+
+    Handle enrollment data and manage the validator set(db)
+
+*******************************************************************************/
+
+public class EnrollmentManager
+{
+    /// SQLite db instance
+    private Database db;
+
+    /// Node's key pair
+    private KeyPair node_key_pair;
+
+    /// Random seed
+    private Scalar random_seed_src;
+
+    /// Path to data file on disk
+    private string node_enroll_data_path;
+
+    /// Random key for enrollment
+    private Pair signature_noise;
+
+    /// Enrollment data object
+    private Enrollment* data;
+
+    /// Database path
+    private string db_path;
+
+    /***************************************************************************
+
+        Constructor
+
+        Params:
+            db_path = path to the database file, or in-memory storage if
+                        :memory: was passed
+            seed_file_path = path to the data file storing randon seed
+            key_pair = the keypair of the owner node
+
+    ***************************************************************************/
+
+    public this (string db_path, string data_dir, KeyPair key_pair)
+    {
+        const db_exists = db_path.exists;
+        if (db_exists)
+            log.info("Loading database from: {}", db_path);
+
+        this.db = Database(db_path);
+
+        // create the table for validator set if it doesn't exist yet
+        this.db.execute("CREATE TABLE IF NOT EXISTS validator_set " ~
+            "(key BLOB PRIMARY KEY, val BLOB NOT NULL)");
+
+        // set node's KeyPair object
+        this.node_key_pair = key_pair;
+
+        // load random seed
+        this.node_enroll_data_path = buildPath(data_dir, "node_enroll_data.dat");
+        loadNodeEnrollmentData();
+
+        // set information to member variables
+        this.db_path = db_path;
+    }
+
+    /***************************************************************************
+
+        Shut down the database
+
+        Note: this method must be called explicitly, and not inside of
+        a destructor.
+
+    ***************************************************************************/
+
+    public void shutdown ()
+    {
+        this.db.close();
+    }
+
+    /***************************************************************************
+
+        Add a enrollment data to the validator set
+
+        Params:
+            enroll = the enrollment data to add
+
+        Returns:
+            true if the enrollment data has been added to the validator set
+
+    ***************************************************************************/
+
+    public bool addEnrollment (Enrollment enroll) @safe
+    {
+        static ubyte[] buffer;
+        buffer.length = 0;
+
+        // check if already exists
+        if (hasEnrollment(enroll.utxo_key))
+        {
+            log.info("Rejected enroll: {}", enroll);
```

Comment:
The message is too broad. Imagine trying to track down a bug in 3 months and seeing this. You at least want to know *why* the enrollment was rejected. Something like "Rejected already existing enrollment: {}" gives much more information at a glance.